Add Problem Response Report Download button

bradenmacdonald opened the pull request:
OpenCraft ticket: [OC-1823](https://tasks.opencraft.com/browse/OC-1823)

**Description**:

With https://github.com/edx/edx-analytics-pipeline/pull/274 , the analytics pipeline gained the ability to generate Problem Response Reports, which are CSV files stored on S3 that contain all responses to all problems by each learner in a course.

With https://github.com/edx/edx-analytics-data-api/pull/131 , the data API gained and endpoint that can be used to retrieve secure, temporary download URLs to access those reports.

Now, with this PR, there is an optional UI in Insights that can be used to download those reports, where available.

This feature is off by default, and can be enabled with the new `enable_problem_response_download` switch.

**Contribution Type**
This is an open source contribution developed by OpenCraft for the Harvard Kennedy School.

**Dependencies**:
* https://github.com/edx/edx-analytics-pipeline/pull/274
* https://github.com/edx/edx-analytics-data-api/pull/131
* https://github.com/edx/edx-analytics-data-api-client/pull/27

**Screenshots**:
New menu item on each course home page:
![screen shot 2016-08-12 at 6 06 17 pm](https://cloud.githubusercontent.com/assets/945577/17640580/85b75798-60b7-11e6-813d-38cea72e273a.png)

Clicking on that opens this new page:
![screen shot 2016-08-12 at 6 32 59 pm](https://cloud.githubusercontent.com/assets/945577/17640727/3df31038-60bb-11e6-8dff-f9801965fcf2.png)


The button will (almost) immediately download the CSV.

If this feature is enabled but the data API reports that the Problem Responses Report is not available for the current course:
![screen shot 2016-08-12 at 6 08 22 pm](https://cloud.githubusercontent.com/assets/945577/17640589/cb817970-60b7-11e6-8833-f55e405e3dc5.png)

**Testing Instructions**
On devstack: 

1. Generate some reports and store them on S3 according to the test instructions in https://github.com/edx/edx-analytics-data-api/pull/131 ; ensure you can go to http://localhost:8100/docs/#!/api/Report_Download and query for the `problem_response` report and a particular course ID to get some data from the API.
2. Apply [**this patch**](https://github.com/open-craft/edx-analytics-data-api/commit/004d1794831544169f2963c62a67a6d0e2f590d3) if it hasn't yet been included in https://github.com/edx/edx-analytics-data-api/pull/131
2. Checkout this branch and update the Insights requirements (`make requirements`). Ensure https://github.com/edx/edx-analytics-data-api-client/pull/27 is installed
3. Run Insights and ensure that this new feature is disabled by default and doesn't appear anywhere in the Performance UI nor on any course home page
4. Enable the switch: `./manage.py waffle_switch enable_problem_response_download on --create`
5. Relaunch Insights and look for the new report links and download button as shown in the screenshots.
6. Test both a course that has CSV reports available, and one that doesn't. Ensure the "last updated" and file size information shown below the download button is correct.
7. Clicking the button should download the CSV.
8. In an incognito window, login to Insights as a user who does not have access to the test course you are using. Ensure that both `/courses/:course_id/performance/problem_responses/` and the direct download URL ("Copy Link Address" from the download button seen by the authorized user) return a permissions error for that user.

**Notes**
* Before merging, fix `requirements/base.txt` to once again refer to `edx-analytics-data-api-client` by version number

pomegranited commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+    <div class="section-content">
+        <p>
+        {% blocktrans %}
+            This downloadable report includes the answer that each learner gave for every supported problem in the course.
```
Could you please change this to: "This downloadable report includes the most recent answer that each learner gave.."

pomegranited commented on analytics_dashboard/courses/views/csv.py:
```diff
@@ -91,3 +92,14 @@ class PerformanceAnswerDistributionCSV(CSVResponseMixin, CourseView):
+    def render_to_response(self, context, **response_kwargs):
+        presenter = CourseReportDownloadPresenter(self.course_id)
+        data = presenter.get_report_info(CourseReportDownloadPresenter.PROBLEM_RESPONSES)
+        return HttpResponseRedirect(data['download_url'], **response_kwargs)
```
This is a lovely way to obscure this URL, and discourage it from getting into unauthenticated hands.  😄

pomegranited commented:
@bradenmacdonald This works as described, thank you!

Other than my text nit, there's just a couple of things failing in the automated tests:
* [pylint on test_performance.py](https://travis-ci.org/edx/edx-analytics-dashboard/builds/151991222#L9017)
* [a11y_tests.test_course_enrollment_demographics_axs.CourseEnrollmentDemographicsAgeTests](https://travis-ci.org/edx/edx-analytics-dashboard/builds/151991222#L9627) - but this looks totally unrelated?  Not sure why they're failing, given that [mine passed just 2 hours earlier](https://travis-ci.org/edx/edx-analytics-dashboard/builds/151977680#L9865).

bradenmacdonald commented:
Thanks @pomegranited! Sorry, I ran out of time on Friday and wasn't able to monitor CI and catch those. The a11y thing is just a flaky test so it will hopefully go away.

Comment + pylint addressed in 9be70cedf6fc927a2dfe231836114428584c0f41 and I've squashed everything down to d18c702bb2e8777b19333dabfbf31b08a9a65eda

pomegranited commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+    <div class="section-content">
+        <p>
+        {% blocktrans %}
+            This downloadable report includes the most recent answer that each learner gave.
```
With my [previous comment](https://github.com/edx/edx-analytics-dashboard/pull/538#discussion_r74684392), I intended just to ask you to add the phrase "the most recent answer" to your existing sentence, making it read: "This downloadable report includes the most recent answer that each learner gave for every supported problem in the course."

My apologies for being unclear!

pomegranited commented:
[One small nit (my mistake!)\(https://github.com/edx/edx-analytics-dashboard/pull/538/files#r74697035), but once that's fixed:

 :+1: 

- [x] I tested this using the given instructions, for courses with an without a report available, and for users with and without instructor access to these courses.
- [x] I read through the code.
- [x] Includes documentation - docstrings and client-facing descriptions are clear.

bradenmacdonald commented:
Thanks @pomegranited! That one issue should be fixed now.

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-1395](https://openedx.atlassian.net/browse/OSPR-1395) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

dsjen commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+
+        {% if report_url %}
+            <div class="section-action">
+                <p><a class="btn btn-default" href="{{ report_url }}" data-role="learner-data-csv" data-track-type="click"
```
For accessibility, we add a `role=button`.  For example, please see https://github.com/edx/edx-analytics-dashboard/blob/master/analytics_dashboard/static/apps/learners/common/templates/download-data.underscore#L3.

dsjen commented on analytics_dashboard/core/templatetags/dashboard_extras.py:
```diff
@@ -103,6 +103,11 @@ def show_metrics_error():
 
+@register.inclusion_tag('section_error.html')
+def show_report_error():
+    return _get_base_error_context('report')
```
Nice!

dsjen commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+
+{% block child_content %}
+
+  <section class="view-section" data-section="enrollment-basics" aria-hidden="true">
```
Is `aria-hidden="true"` needed here?

bradenmacdonald replied:
I copied the template from [other pages](https://github.com/edx/edx-analytics-dashboard/search?utf8=%E2%9C%93&q=data-section%3D%22enrollment-basics%22+aria-hidden&type=Code), but I guess the intention there is to hide the whole data graph section from audio-only users. I will remove `aria-hidden="true"` here.

dsjen commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+            <div class="section-action">
+                <p><a class="btn btn-default" href="{{ report_url }}" data-role="learner-data-csv" data-track-type="click"
+                   data-track-event="edx.bi.csv.downloaded" data-track-category="problem-response-report">
+                  <span class="ico fa fa-download"></span>
```
An `aria-hidden="true"` would keep the icon hidden from screen readers. :)

dsjen commented on analytics_dashboard/courses/tests/test_views/__init__.py:
```diff
@@ -96,6 +96,9 @@ def get_mock_data(self, course_id):
 # pylint: disable=not-callable,abstract-method
 @ddt
 class AuthTestMixin(MockApiTestMixin, PermissionsTestMixin, RedirectTestCaseMixin, UserTestCaseMixin):
+    follow = True  # Should test_authentication() and test_authorization() follow redirects
```
Thank you!

dsjen commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+        {% if report_url %}
+            <div class="section-action">
+                <p><a class="btn btn-default" href="{{ report_url }}" data-role="learner-data-csv" data-track-type="click"
+                   data-track-event="edx.bi.csv.downloaded" data-track-category="problem-response-report">
```
It looks like we're using underscores instead of hyphens for the track categories (e.g. problem_response_report).  For example, https://github.com/edx/edx-analytics-dashboard/blob/f770469e164b5c2bbe5a27a255b3feee5032a4ed/analytics_dashboard/static/apps/learners/common/views/download-data.js#L24.

bradenmacdonald replied:
Ah, thanks. Changed to `problem_response_report`.

dsjen commented on analytics_dashboard/courses/templates/courses/performance_problem_responses.html:
```diff
@@ -0,0 +1,40 @@
+                {% if report_download_size_kb %}
+                  {% blocktrans %}Download size: {{report_download_size_kb}} KiB{% endblocktrans %}
+                {% endif %}
+                <p>
```
This looks like an extra `<p>` to me?

bradenmacdonald replied:
Me too. Removed.

dsjen commented:
@edx/edx-accessibility -- this is an OSPR that includes some HTML templates.  I gave it an a11y pass and made a few recommendations.  Please let us know if I've missed anything.

dsjen commented:
@bradenmacdonald -- thanks for putting this behind a switch!  Would you please also add documentation for it under the [feature gating section](https://github.com/edx/edx-analytics-dashboard/blob/master/README.md#feature-gating) too?

dsjen commented:
@bradenmacdonald -- @alyssaboehm & I chatted about the UX around this functionality and her suggestion would be reduce the number of clicks needed to get to the CSV download.  I've included a screenshot of an in-page mockup that includes a change to the icon and additional text to describe what you're getting when you click on the link.  Clicking the link would then download the CSV.  Please let us know what you think.

![screen shot 2016-09-09 at 9 49 20 am](https://cloud.githubusercontent.com/assets/5265058/18389205/dab37d3c-7672-11e6-9a4d-c91c452d6e20.png)

dsjen commented:
@bradenmacdonald -- just wanted to check in and see if these changes made sense.  Please let us know if you need anything from us!

bradenmacdonald commented:
@dsjen Sorry for the delay in addressing your review comments. I've addressed the minor comments with eb8e4bf94d03f0d6306cf050496c4a405fbc9e39.

Regarding that mockup: It's true that the current approach requires more clicks before the user can download the CSV, but it has the advantage that it gives us space to properly explain what the data download will contain, and to advise the user about the size of the download.

One issue with the approach shown in the mockup is that it indicates that problem responses are part of the "Performance" tab, but there would be no way to download them from within the performance tab (i.e. If I click on "Ungraded Problems" first), unless we also keep the new page in.

Which would you like me to use? I don't have a strong opinion so I'm happy to change to what's shown in that mockup.

Also, is it ok if I rebase this now?

dsjen commented:
@bradenmacdonald -- thank you!  Please feel free to rebase.

@alyssaboehm -- do you have thoughts about Braden's concerns?  My sense is that fewer clicks is better in this case and Braden's point about the having clicked on the performance tab and not seeing and download does hide it a bit.

Thanks, all!

bradenmacdonald commented:
Rebased from eb8e4bf94d03f0d6306cf050496c4a405fbc9e39 to 651a5d1118be1c640cdf0614cd393dffe669bd7d.

Just waiting for input about the UX.

dsjen commented:
Thanks, @bradenmacdonald!  @alyssaboehm and I chatted and think that it would be better to shift the download to the course landing page rather than add another page.  The reasoning being that the download size doesn't add much value unless the download is extremely large and the navigation for the pages is undergoing redesign.

Sorry that you built out the download page, and I hope it's not a pain to switch up the design.  Please let us know if there's anything we can do to help.

bradenmacdonald commented:
@dsjen Sure thing. Sorry for the slow response here. I'm hoping to have this updated to match that new UX by my end of day today.

dsjen commented:
Thank you!

bradenmacdonald commented:
@dsjen @alyssaboehm I noticed one issue with this mockup:
![screen shot 2016-09-09 at 9 49 20 am](https://cloud.githubusercontent.com/assets/5265058/18389205/dab37d3c-7672-11e6-9a4d-c91c452d6e20.png)

The text below each link is supposed to be navigation breadcrumbs, and so the icon for each breadcrumb in the performance tab is `fa-check-square-o` to match the header above - in other words, the breadcrumbs are shorthand for things like "Performance > Ungraded Problems". But in the mockup, it shows "(Download icon) > Problem Response Report (CSV Download)" which is inconsistent with the other breadcrumbs. What do you think about putting the download icon where it says (CSV Download), or removing that icon completely?

![screen shot 2016-09-19 at 5 24 45 pm](https://cloud.githubusercontent.com/assets/945577/18653329/005b46d2-7e8e-11e6-99d2-d39c4961711a.png)

bradenmacdonald commented:
@dsjen I've pushed a commit that makes those changes.

Because of the open question above, here's what I've implemented so far (different icon):
![screen shot 2016-09-19 at 6 21 13 pm](https://cloud.githubusercontent.com/assets/945577/18654424/ea7c5a60-7e95-11e6-973e-4d765d934ea8.png)

One issue to be aware of with this approach is that if the report is unavailable, the user would get an ugly 404 error from clicking that link. So to avoid that, the course home page has to make an API call to determine if the report is actually available for the current course. There's also no way to tell the difference between "problem response report is not enabled" and "problem response report is enabled but no data is available." Let me know your thoughts on this.

bradenmacdonald commented:
Build errors seem unrelated (a11y failures on other parts of the codebase)

dsjen commented:
@bradenmacdonald -- thanks for the update!  I realize we don't have infrastructure to adjust the styling of the lens breadcrumbs easily and personally, I think the second mockup (with the download icon within the parentheses) suits me.

@alyssaboehm -- what do you think?

alyssaboehm commented:
Agreed!

On Tue, Sep 20, 2016 at 3:30 PM, Dennis Jen <notifications@github.com>
wrote:

> @bradenmacdonald <https://github.com/bradenmacdonald> -- thanks for the
> update! I realize we don't have infrastructure to adjust the styling of the
> lens breadcrumbs easily and personally, I think the second mockup (with the
> download icon within the parentheses) suits me.
>
> @alyssaboehm <https://github.com/alyssaboehm> -- what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-analytics-dashboard/pull/538#issuecomment-248407512>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATf3aLRo2NYz0O2T9vz3tZRirO_oLhvvks5qsDRDgaJpZM4JjkqW>
> .
>

bradenmacdonald commented:
@dsjen Done! (via edb58d7bcca08c426d503e3f793845ab7959607b)

No new screenshot because it looks exactly like the mockup.

thallada commented:
This looks great :+1: Thanks for making that UX change.

dsjen commented:
@bradenmacdonald -- good to go!  👍  🐹  💯 

Please squash commits and feel free to merge it in (you have the right permissions, right?).

bradenmacdonald commented:
Merged. Thanks @dsjen @alyssaboehm @thallada @pomegranited  !

I'm about to leave on vacation, so please contact @pomegranited if there's any post-merge follow up required.

dsjen commented:
Thank you for being so patient with us and have nice break!
